### ISSUE TITLE
Running share-mnt binary manually results in trace/breakpoint trap

### ISSUE THREAD

**Issue by sukrit007** [rancher/rancher#20740]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- Configure the rancher server
- Start the rancher agent (v2.2.x)
- Rancher provisioning fails due to issue in share-mnt binary

**Result:**

**Other details that may be helpful:**
Rancher 2.2.x includes https://github.com/rancher/share-mnt/releases/download/v1.0.4/share-mnt-amd64.tar.gz .  This binary on execution results in segmentation fault on AMD64 architecture.

```
docker run --privileged --net host --pid host -v /:/host --rm -it  --entrypoint /bin/bash  rancher/rancher-agent:v2.2.3    
root@linuxkit-025000000001:/# share-mnt
Trace/breakpoint trap
root@linuxkit-025000000001:/#
```

However,  the agent with 1.0.3 version of share-mnt works.
```
docker run --privileged --net host --pid host -v /:/host --rm -it  --entrypoint /bin/bash  rancher/rancher-agent:v2.1.10   
root@linuxkit-025000000001:/# share-mnt
Found container ID: 3cf22ef2d0f4367128a6f3f69ebed139c52279e57b15227a6fbf8f2d64b147d4
Checking root: /host/run/runc
Checking root: /host/var/run/runc
Checking root: /host/run/docker/execdriver/native
Checking root: /host/var/run/docker/execdriver/native
Checking root: /host/run/docker/runtime-runc/moby
Checking root: /host/var/run/docker/runtime-runc/moby

```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.2.3
- Installation option (single install/HA): single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): Custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): cloud + VM
- Kubernetes version (use `kubectl version`): 1.14

- Docker version (use `docker version`):

```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:39 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false

```

**Comment by sukrit007** [comment]
Untill this issue gets fixed,  I have patched agent with 1.0.3 version of share-mnt binary. 

See https://github.com/intraedge-services/rancher-agent-patch

**Comment by superseb** [comment]
Can you share the full log of the `share-mnt` container when it is run when you add a node to the cluster? Also what OS and kernel is this on?

**Comment by sukrit007** [comment]
@superseb Here are the actual logs from 'share-mnt' container that got created when rancher was provisioned.  (All the sensitive info is removed)
```
INFO: Arguments: -- share-root.sh docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/
var/run <registry-url>/rancher/rancher-agent:v2.2.3 --server https://<rancher-server> --token REDACTED --ca-checksum <ca-checksum>
<token> --no-register --only-write-certs /var/lib/kubelet /var/lib/rancher
+ trap 'exit 0' SIGTERM
++ grep :devices: /proc/self/cgroup
++ head -n1
++ awk -F/ '{print $NF}'
++ sed -e 's/docker-\(.*\)\.scope/\1/'
+ ID=d87baf20aeffcb44b62a1364c2ff2ae496c00202b136ee71e895394b32808404
++ docker inspect -f '{{.Config.Image}}' d87baf20aeffcb44b62a1364c2ff2ae496c00202b136ee71e895394b32808404
+ IMAGE=<registry-url>/rancher/rancher-agent:v2.2.3
+ bash -c 'docker run -d --privileged --restart=unless-stopped --net=host -v /etc/kubernetes:/etc/kubernetes -v /var/run:/var/run dev.registry.ew
s.int:443/vendored/rancher/rancher-agent:v2.2.3 --server https://<rancher-server> --token <token> --ca-ch
ecksum <ca-checksum> --no-register --only-write-certs'
4d8f955b5a73eacc2d753ea94fdb9a5cf1732aae6314a4b54d67855321637e21
+ docker run --privileged --net host --pid host -v /:/host --rm --entrypoint /usr/bin/share-mnt <registry-url>/rancher/rancher
-agent:v2.2.3 /var/lib/kubelet /var/lib/rancher -- norun
+ docker start kubelet
Error response from daemon: {"message":"No such container: kubelet"}
Error: failed to start containers: kubelet
+ sleep 2
+ docker start kubelet
Error response from daemon: {"message":"No such container: kubelet"}
Error: failed to start containers: kubelet
........
Above error repeats forever
........
```

From the above log, I tried running following command on different flavor of VMs (amd64 architecture), and it was failing with exit code 133  (Segmentation Fault) . The segmentation fault error is not shown in the log,  but after I ran a bash 
```
[user@<host> ~]$ sudo docker run --privileged --net host --pid host -v /:/host --rm --entrypoint /usr/bin/share-mnt <registry-url>/rancher/rancher-agent:v2.2.3 /var/lib/kubelet /var/lib/rancher -- norun
[user@<host> ~]$ echo $?
133
[user@<host> ~]$ sudo docker run --privileged --net host --pid host -v /:/host -it --rm --entrypoint /bin/bash <registry-url>/rancher/rancher-agent:v2.2.3
user@<host>:/# /usr/bin/share-mnt /var/lib/kubelet /var/lib/rancher -- norun
Trace/breakpoint trap (core dumped)
```

OS and kernel version
```
CentOS Linux release 7.6.1810 (Core)
[user@host~]$ uname -a
Linux s8497dalv 3.10.0-957.10.1.el7.x86_64 #1 SMP Mon Mar 18 15:06:45 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[user@host~]$
```

Once I found the issue,  I tried running container on Mac, Ubuntu 16 and I run into same segmentation fault.  After I download share-mnt binary (1.0.3) within the container,  the error goes away 
```
curl -sL https://github.com/rancher/share-mnt/releases/download/v1.0.3/share-mnt.tar.gz | tar xvzf - -C /usr/bin/
```

I also tried to build and run share-mnt on my Mac , it had same issue.  I got the binary to work by removing external linkage. This works fine on alpine container,  however for some reason breaks on Ubuntu
See https://github.com/rancher/share-mnt/pull/9

**Comment by sukrit007** [comment]
**Update on the issue:**   
I was able to get around share-mnt by re-creating the cluster. The binary still has the issue,  but I imagine the step is optional.

However,  after I cleaned up the containers (not the data) and started agent again, it again got stuck on share-mnt issue.

**Comment by superseb** [comment]
Yes, it doesn't block the provisioning from succeeding as it works as it should when it's run as it should. We can look into why the error occurs when the binary is ran manually though.

For the actual issue, please provide the steps from the issue template and the logs so we can use that to reproduce. Setting up a custom cluster and adding a node doesn't trigger this, so we need a bit more info to narrow down possible causes. Which is basically https://github.com/rancher/rancher/issues/13314 but can't tell for sure without the requested info.

**Comment by gknepper** [comment]
I tried to apply this fix but unfortunately it's not working for me, my installation still stuck in a endless loop of

`+ docker start kubelet
Error response from daemon: {"message":"No such container: kubelet"}
Error: failed to start containers: kubelet
+ sleep 2
`

**Comment by superseb** [comment]
To reproduce:

1. Create custom cluster and add nodes
2. Check if kernel traps message is logged in system logging (`/var/log/syslog`, `/var/log/messages`, `/var/log/kern.log`) using `grep traps $FILE`
```
kernel: [  108.318555] traps: share-mnt[6515] trap int3 ip:4546f2 sp:7ffd8cf26a58 error:0 in share-mnt[400000+385000]
```
3. Run share-mnt binary manually and check exit code:

```
docker run --entrypoint share-mnt $(docker inspect $(docker images -q --filter=label=io.cattle.agent=true) --format='{{index .RepoTags 0}}' | tail -1); echo $?
133
```
To validate:

1. Create custom cluster and add nodes
2. kernel traps message should not be seen
3. Running the share-mnt binary should actually start and show logging, and fail with `Failed to find state.json` and exit with code 1 and not 133.

**Comment by jiaqiluo** [comment]
Reproduced in the custom cluster node in Rancher: v2.3.3

Steps:
- run Rancher:v2.3.3 single install
- add a custom cluster 

Results:
- ssh into one of the nodes, the logs in the node show the following errors
```
root@jiaqi-tf--rancheragent-0-all:~# cat /var/log/kern.log | grep trap
Dec 13 22:07:03 jiaqi-tf--rancheragent-0-all kernel: [  106.031038] traps: share-mnt[5966] trap int3 ip:4546f2 sp:7ffdad373ef8 error:0 in share-mnt[400000+385000]
root@jiaqi-tf--rancheragent-0-all:~# cat /var/log/syslog | grep trap
Dec 13 22:07:03 jiaqi-tf--rancheragent-0-all kernel: [  106.031038] traps: share-mnt[5966] trap int3 ip:4546f2 sp:7ffdad373ef8 error:0 in share-mnt[400000+385000]
root@jiaqi-tf--rancheragent-0-all:~#
```

- Run share-mnt binary manually and check exit code
```
root@jiaqi-tf--rancheragent-0-all:~# docker run --entrypoint share-mnt $(docker inspect $(docker images -q --filter=label=io.cattle.agent=true) --format='{{index .RepoTags 0}}' | tail -1); echo $?
133
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in v2.3-head 9626c3f9e

Results:
- no error in the logs in the node 
- run share-mnt binary manually and the exit code is 1 
```
root@jiaqi-tf--rancheragent-1-all:~# docker run --entrypoint share-mnt $(docker inspect $(docker images -q --filter=label=io.cattle.agent=true) --format='{{index .RepoTags 0}}' | tail -1); echo $?
Found container ID: ca6fb582bf511282cdda82756415b177fdb30c03ab01c17b6cf84ea1d4908a63
Checking root: /host/run/runc
Checking root: /host/var/run/runc
Checking root: /host/run/docker/execdriver/native
Checking root: /host/var/run/docker/execdriver/native
Checking root: /host/run/docker/runtime-runc/moby
Checking root: /host/var/run/docker/runtime-runc/moby
Checking root: /host/run/docker/runtime-nvidia/moby
Checking root: /host/var/run/docker/runtime-nvidia/moby
time="2019-12-13T22:37:15Z" level=fatal msg="Failed to find state.json"
1
```

**Comment by jiaqiluo** [comment]
The bug fix is validated in master-head 9e5545525

Results:
- no error in the logs in the node 
- run share-mnt binary manually and the exit code is 1 
```
root@jiaqi-tf--rancheragent-2-all:~#  docker run --entrypoint share-mnt $(docker inspect $(docker images -q --filter=label=io.cattle.agent=true) --format='{{index .RepoTags 0}}' | tail -1); echo $?
Found container ID: 6ea476646b67d37d0414b0cf5ed889b6589d7903a64ac484ec97ccac2d42f586
Checking root: /host/run/runc
Checking root: /host/var/run/runc
Checking root: /host/run/docker/execdriver/native
Checking root: /host/var/run/docker/execdriver/native
Checking root: /host/run/docker/runtime-runc/moby
Checking root: /host/var/run/docker/runtime-runc/moby
Checking root: /host/run/docker/runtime-nvidia/moby
Checking root: /host/var/run/docker/runtime-nvidia/moby
time="2019-12-13T22:50:34Z" level=fatal msg="Failed to find state.json"
1
```
